ledger-service/db-backend/src/main/scala/com/digitalasset/http/dbbackend/Queries.scala

```diff
@@ -303,24 +304,31 @@ sealed abstract class Queries(tablePrefix: String, tpIdCacheMaxEntries: Long)(im
       dbcs: F[DBContract[SurrogateTpId, DBContractKey, JsValue, Array[String]]]
   )(implicit log: LogHandler): ConnectionIO[Int]
 
+  // ContractTypeId -> CId[String]
   final def deleteContracts(
-      cids: Set[String]
+      cids: Map[SurrogateTpId, Set[String]]
   )(implicit log: LogHandler): ConnectionIO[Int] = {
-    import cats.data.NonEmptyVector
     import cats.instances.vector._
     import cats.instances.int._
     import cats.syntax.foldable._
-    NonEmptyVector.fromVector(cids.toVector) match {
-      case None =>
+    import nonempty.catsinstances._
+    (cids: Iterable[(SurrogateTpId, Set[String])]).view.collect { case (k, NonEmpty(cids)) =>
+      (k, cids)
+    }.toMap match {
+      case NonEmpty(cids) =>
+        val chunks = maxListSize.fold(Vector(cids))(chunkBySetSize(_, cids))
```

Comment:
If I understand correctly the maxListSize limits how many rows are selected for the delete at a time?

Reply:
It's probably overconservative to make the chunks as small as they are now, since the limiting factor seems to be the `in`. Maybe we should just go back to a single query that joins everything with `OR` all at once, only using chunks to limit the size of each argument passed to `in`.

Reply:
#15412